### ISSUE TITLE
What's in HTMLy 3.0?

### ISSUE THREAD

**Issue by thopanX** [danpros/htmly#718]
This is more of a question: Are there any big(ger) features in the pipeline for HTMLy for 3.0? What does the maintainer have in mind for the project?

Since installing it again, I've been delighted by the uptake in releases and the responsiveness of the maintainer.

**Comment by danpros** [comment]
There will be no major changes, the main focus remains a platform for easy and simple blogging. Of course, performance, security, and compatibility remains a priority too. Until PHP 8.3, there was no deprecated warning or error message so it can be considered ready for any PHP 8.x version or even PHP 9?

And we will keep the structure as it is now (the most important metadata is in the folder and file names). As a platform without a database, data integrity is very important, so we try not to store all the metadata (category, tags etc.) in one file, because if this file is corrupted, of course you can imagine the problems that will arise.

Do you have any input? I would love to hear that.

**Comment by CharliePoole** [comment]
Agree about data integrity. Because I've had to move my installations around several times, I've seen a few problems, which can arise. Here is one thing I found, which doesn't happen in normal use, but can when copying files...

* If the menu.json file disappears, there doesn't seem to be any way to escape from auto-generated mode except to create one manually, as I did. Maybe the Admin/Menu page could have a check or button to switch to custom mode. You would have to decide what the initial menu would look like... maybe just Home or Home and Blog.

For new features, I'd love to see some sort of layout setting for pages - an extension of issue #717 - giving pages the same level of flexibility which exists for posts with categories.

I'd also like to see a bit more customization available for the secondary column, in case I want to populate it with something other than the standard content. This could be a template file included in the display in the same way that the primary column is generated.

I would love to see Mastodon supported as a standard social link at least. If it can be used as a commenting system as @thopanX suggests, that would be even better.

As you see, I'm focusing a lot on pages rather than posts. That's cause I'm pretty happy with how it works for blogging but I'd like to do more with pages on my own site.

**Comment by rolandixor** [comment]
Please, please, please :)... add autosave support, or at least, request to save before navigating away from a draft. I lost an important post draft (nearly two) on my first time trying it out.

Really love the blog engine as it serves my needs quite well, and I'd recommend it to others, but this could be a big problem.

**Comment by vdbhb59** [comment]
If asking to add support for commenting, I may have a solution which my Spouse made for me, when around - 1.5-2 years ago HTMLy was in hiatus. But I am not too sure if it works fully, and may need some tweaking. However, on that note, commenting should remain within HTMLy as a guest/loggedin-user as well, and not with a 3rd party. Which is what I was aiming for. Flat file CMS should allow for that, and in earlier days, that is how users without any login used to comment.

**Comment by Clover-Zero** [comment]
* I'd like a focus mode as well. It doesn't have to be full-screen, but I want the editor to be free from clutter, something like WriteFreely. Bonus points if there's a dark theme. That'd mean no real-time preview, but that's fine for me.
* I'd like ActivityPub support as well. At least something like how WriteFreely works.
* Maybe support for using [Cusdis](https://cusdis.com/) for comment system?

**Comment by KuJoe** [comment]
> * I'd like a focus mode as well. It doesn't have to be full-screen, but I want the editor to be free from clutter, something like WriteFreely. Bonus points if there's a dark theme. That'd mean no real-time preview, but that's fine for me.
> 
> * I'd like ActivityPub support as well. At least something like how WriteFreely works.
> 
> * Another vote for auto-save too.
> 
> * Maybe support for using [Cusdis](https://cusdis.com/) for comment system?

I like the idea of a focus mode, I'll definitely look into this.

ActivityPub support would be great, but I don't know how easy it would be to implement with HTMLy natively while keeping it optional. Maybe a plugin/addon would work best like how Wordpress handles it. I'll definitely do more research into it as my only exposure to ActivityPub is Mastodon.

Auto-save has been added and will be in a future release.

It seems the comment systems people want vary quite a bit so maybe I'll look into if there's a better way to implement them to allow for a wider range of options without having to hardcode any of them into the configs. Again, maybe a module/plugin/addon system to give people options.

**Comment by danpros** [comment]
For focus mode this tips perhaps can help a little #560

Plugin/addon in htmly? all features should built-in like current htmly. From one of htmly user: "I need a tool; not a toolbox".

Like most other flat files, we have to be willing to get our hands dirty if we want to customize some parts of it (customizing the theme is more than enough). Example for comment integration, usually we just need to edit 2 files, `layout.html.php` (the layout, contain head section etc.) and `post.html.php` (individual post), yes just copy and paste a few lines of codes provided by them and done.

There must be a compromise on what features are suitable/good for a simple flat-file blog like htmly.

**Comment by vdbhb59** [comment]
True. Keep it simple and stick to what HTMLy is. Too much makes for a bad apple.

**Comment by rolandixor** [comment]
I don't think plugins would be a bad idea. I would think they provide the perfect compromise.

**Comment by KuJoe** [comment]
> Plugin/addon in htmly? all features should built-in like current htmly. From one of htmly user: "I need a tool; not a toolbox".
>

Understood! Killing the plugin/addon idea. 🙂

**Comment by rolandixor** [comment]
> > Plugin/addon in htmly? all features should built-in like current htmly. From one of htmly user: "I need a tool; not a toolbox".
> 
> Understood! Killing the plugin/addon idea. 🙂

I think having the *option* for plugins would be great. No need to ship them, just allow them.

**Comment by splifingate** [comment]
The idea that there is One Software, to fulfill All Our Needs, is attractive.

Having (truly) dived-deep into the HTMLy code, I am continually appreciative of the simplicity.

The complexity of said simplicity needs no additional ;)

**Comment by KuJoe** [comment]
Would something like this be acceptable for a focus mode? I've added a button to toggle between the normal view and this view:
![image](https://github.com/danpros/htmly/assets/1040086/af1197eb-269d-465f-91af-8bae8acfbc45)

**Comment by vdbhb59** [comment]
Looks alright, with minimal changes and low coding. :)

**Comment by bttrx** [comment]
@danpros Maybe you could integrate the [yellow-comment][1] plugin for use as a simple commenting system. It's under GNU GPL v2.

[1]: https://github.com/GiovanniSalmeri/yellow-comment "GiovanniSalmeri/yellow-comment: Easy commenting system."

**Comment by danpros** [comment]
@bttrx Thanks, I will look into it.
